File: Samples/Canary/CSharp/GSharp.PythonRegressionAnalysis/GSharp.PythonRegressionAnalysis/LinearRegression.cs
using GHost.Core;
using GHost.Core.CanaryLabs.Contracts.Enums;
using GHost.Core.CanaryLabs.Contracts.Model;
using GHost.Core.Logging.Context;
using GHost.Core.Time;
using GHost.GSharp.Calculation.CanaryLabs;
using GHost.GSharp.Core.Attributes;
using GHost.GSharp.Core.CanaryLabs.Data;
using GHost.GSharp.Core.Enums;
using GHost.GSharp.Core.Interfaces.Model;
using GHost.GSharp.PythonNet.Calculation;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Threading;

namespace GSharp.PythonRegressionAnalysis;

/// <summary>
/// A gSharp execution module.
/// This module will be imported to the server at the time the containing assembly is registered.
/// </summary>
public class LinearRegression : CalculationModuleCLBase
{
    private PythonHost _PythonHost;

    [PointDataDirectionInput]
    public CLDoublePoint Well_Level { get; set; }

    [PointDataDirectionOutput]
    public CLDoublePoint Intercept { get; set; }

    [PointDataDirectionOutput]
    public CLDoublePoint Slope { get; set; }


    public override void Initialize()
    {
        _PythonHost = this.InitializePython();
    }

    public override void Execute(ISchedule schedule, DateTime timestamp, CancellationToken cancelToken)
    {
        try
        {
            var tsDataset = Well_Level.Data.GetAggregatedData("Now-8h", "Now", AggregateType.Average, TimeSpan.FromMinutes(5));

            // Clean the data before further processing, i.e., remove any bad quality values.
            var cleanDataSet = tsDataset.Where(v => v.Q >= 64).ToList();

            var pyDictionary = new JsonDataDictionary(cleanDataSet);
            string pyDataset = JsonSerializer.Serialize(pyDictionary);

            string code = ReadPythonScript("LinearRegression.py");

            object pySlope;
            object pyIntercept;

            using (ModuleScope scope = _PythonHost.CreateScope())
            {
                scope.Set("data", pyDataset);
                scope.Exec(code);
                scope.TryGet<object>("slope", out pySlope);
                scope.TryGet<object>("intercept", out pyIntercept);
            }

            double.TryParse(pySlope.ToString(), out double slopeValue);
            double.TryParse(pyIntercept.ToString(), out double interceptValue);

            Slope.Value = slopeValue;
            Intercept.Value = interceptValue;
        }
        catch (Exception ex)
        {
            Log.Error($"An error occurred during execution: {ex}");
        }
    }

    private string ReadPythonScript(string fileName)
    {
        var assembly = System.Reflection.Assembly.GetExecutingAssembly();
        string resourcePath = fileName;
        resourcePath = assembly.GetManifestResourceNames()
            .Single(str => str.EndsWith(fileName));

        using Stream stream = assembly.GetManifestResourceStream(resourcePath);
        using StreamReader reader = new(stream);
        return reader.ReadToEnd();
    }

    //private string ReadPythonScript(string fileName)
    //{
    //    var assembly = System.Reflection.Assembly.GetExecutingAssembly();
    //    string filePath = Path.Combine(Path.GetDirectoryName(assembly.Location), "PythonScripts", fileName);

    //    using Stream stream = File.OpenRead(filePath);
    //    using StreamReader reader = new(stream);
    //    return reader.ReadToEnd();
    //}

    public override void Dispose()
    {
        // Place any class-level disposal code in here.
        // This method is called only once, at the time the context is unscheduled or disabled.
    }
}

internal class JsonDataDictionary
{
    public List<string> Date { get; private set; }

    public List<string> Value { get; private set; }

    public JsonDataDictionary(List<TagValue> data)
    {
        Date = [];
        Value = [];

        for (int i = 0; i < data.Count -1; i++)
        {
            Date.Add(data[i].T.ToString("yyyy-MM-ddTHH:mm:ss"));
            Value.Add(data[i].V.ToString());
        }
    }
}

File: Samples/Canary/VB.NET/GSharp.PythonRegressionAnalysis/GSharp.PythonRegressionAnalysis/LinearRegression.vb
Imports System.IO
Imports System.Reflection
Imports System.Text.Json
Imports System.Threading
Imports GHost.Core.CanaryLabs.Contracts.Enums
Imports GHost.Core.CanaryLabs.Contracts.Model
Imports GHost.GSharp.Calculation.CanaryLabs
Imports GHost.GSharp.Core.Attributes
Imports GHost.GSharp.Core.CanaryLabs.Data
Imports GHost.GSharp.Core.Interfaces.Model
Imports GHost.GSharp.PythonNet.Calculation
Imports Python.Runtime

''' <summary>
''' A gSharp execution module.
''' This module will be imported to the server at the time the containing assembly is registered.
''' </summary>
Public Class LinearRegression
    Inherits CalculationModuleCLBase

    Public _PythonHost As PythonHost

    <PointDataDirectionInput>
    Public Property Well_Level As CLDoublePoint

    <PointDataDirectionOutput>
    Public Property Intercept As CLDoublePoint

    <PointDataDirectionOutput>
    Public Property Slope As CLDoublePoint


    Public Overrides Sub Initialize()
        _PythonHost = InitializePython()
    End Sub

    Public Overrides Sub Execute(schedule As ISchedule, timestamp As Date, cancelToken As CancellationToken)
        Try
            Dim tsDataset = Well_Level.Data.GetAggregatedData("Now-8h", "Now", AggregateType.Average, TimeSpan.FromMinutes(5))

            ' Clean the data before further processing, i.e., remove any bad quality values.
            Dim cleanDataSet = tsDataset.Where(Function(v) v.Q >= 64).ToList()

            Dim pyDictionary = New JsonDataDictionary(cleanDataSet)
            Dim pyDataset = JsonSerializer.Serialize(pyDictionary)
            Dim code As String = ReadPythonScript("LinearRegression.py")

            Dim pySlope As Object = Nothing
            Dim pyIntercept As Object = Nothing

            Using scope As ModuleScope = _PythonHost.CreateScope()
                Dim pyInputData As PyObject = pyDataset.ToPython()
                scope.Set("data", pyInputData)
                scope.Exec(code)
                scope.TryGet("slope", pySlope)
                scope.TryGet("intercept", pyIntercept)
            End Using

            Dim slopeValue As Single = Convert.ToDouble(pySlope)
            Dim interceptValue As Single = Convert.ToDouble(pyIntercept)

            Intercept.Value = interceptValue
            Slope.Value = slopeValue

        Catch ex As Exception

        End Try
    End Sub

    Private Function ReadPythonScript(fileName As String) As String
        Dim assembly As Assembly = assembly.GetExecutingAssembly()
        Dim resourcePath As String = fileName

        resourcePath = assembly.GetManifestResourceNames().Single(Function(str) str.EndsWith(fileName))
        Using stream As Stream = assembly.GetManifestResourceStream(resourcePath)
            Using reader As New StreamReader(stream)
                Return reader.ReadToEnd()
            End Using
        End Using
    End Function

    ' An alternative method for reading the Python script. In this case the script
    ' file would not be an embedded resource, but copied to the output directory on build,
    ' and then deployed with the gSharp assembly and it's other dependencies.
    ' An advantage of this approach is that the Python script could be updated as
    ' needed without recompiling the gSharp calculation assembly.

    'Private Function ReadPythonScript(fileName As String) As String
    '    Dim assembly As Assembly = Assembly.GetExecutingAssembly()
    '    Dim filePath As String = Path.Combine(Path.GetDirectoryName(assembly.Location), "PythonScripts", fileName)

    '    Using stream As Stream = File.OpenRead(filePath)
    '        Using reader As New StreamReader(stream)
    '            Return reader.ReadToEnd()
    '        End Using
    '    End Using
    'End Function

    Public Overrides Sub Dispose()
        ' Place any class-level disposal code in here.
        ' This method is called only once, at the time the context is unscheduled Or disabled.
    End Sub
End Class

Friend Class JsonDataDictionary
    Public Property [Date] As List(Of String)
    Public Property Value As List(Of String)

    Public Sub New(data As List(Of TagValue))
        [Date] = New List(Of String)
        Value = New List(Of String)

        For i As Integer = 1 To data.Count
            [Date].Add(data(i).T.ToString("yyyy-MM-ddTHH:mm:ss"))
            Me.Value.Add(data(i).V.ToString())
        Next
    End Sub
End Class